functional/generics-type-classes.md

```diff
@@ -0,0 +1,215 @@
+    use super::{bootp, nfs};
+    pub(crate) trait ProtoKind {
+        type AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo;
+    }
+    pub struct Nfs(nfs::AuthInfo, PathBuf); // the mount point metadata
+    impl Nfs {
+        pub(crate) fn mount_point(&self) -> &Path {
+            &self.1
+        }
+    }
+    impl ProtoKind for Nfs {
+        type AuthInfo = nfs::AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo {
+            self.0.clone()
+        }
+    }
+    pub struct Bootp(); // no additional metadata
+    impl ProtoKind for Bootp {
+        type AuthInfo = bootp::AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo {
+            bootp::AuthInfo()
+        }
+    }
+}
+
+use proto_trait::ProtoKind; // keep internal to prevent impls
+pub use proto_trait::{Nfs, Bootp}; // re-export so callers can see them
+
+struct FileDownloadRequest<P: ProtoKind> {
+    file_name: PathBuf,
+    protocol: P,
+}
+
+// all common API parts go into a generic impl block
+impl<P: ProtoKind> FileDownloadRequest<P> {
+    fn file_path(&self) -> &Path {
+        &self.file_name
+    }
+
+    fn auth_info(&self) -> P::AuthInfo {
+        self.protocol.auth_info()
+    }
+}
+
+// all protocol-specific impls go into their own block
+impl FileDownloadRequest<Nfs> {
+    fn mount_point(&self) -> &Path {
+        self.protocol.mount_point()
+    }
+}
+```
+
+With this approach, if the user were to make a mistake and use the wrong
+type;
+
+```rust,ignore
+fn main() {
+    let mut socket = crate::bootp::listen()?;
+    while let Some(request) = socket.next_request()? {
+        match request.mount_point().as_ref()
+            "/secure" => socket.send("Access denied"),
+            _ => {} // continue on...
+        }
+        // Rest of the code here
+    }
+}
+```
+
+They would get a syntax error. The type `FileDownloadRequest<Bootp>` does not
+implement `mount_point()`, only the type `FileDownloadRequest<Nfs>` does. And
+that is created by the NFS module, not the BOOTP module of course!
+
+## Advantages
+
+First, it allows fields that are common to multiple states to be de-duplicated.
+By making the non-shared fields generic, they are implemented once.
+
+Second, it makes the `impl` blocks easier to read, because they are broken down
+by state. Methods common to all states are typed once in one block, and methods
+unique to one state are in a separate block.
+
+Both of these mean there is less code and it is better organized.
+
+## Disadvantages
+
+This increases code size. Hopefully the compiler will be more intelligent about
+the way monomorphization is implemented in the future.
+
+## Alternatives
+
+If a type seems to need a "split API" due to construction or partial
+initialization, consider the
+[Builder Pattern](../patterns/creational/builder.md) instead.
+
+If the split API is for methods in different state of a state machine, consider
+the [Strategy Pattern](../patterns/behavioural/strategy.md) instead.
```

Comment:
I think it would be good to list type state as an alternative, either here or there we should have a comparison to let others know the differences between these two.